### PR TITLE
JW-450_flows_device

### PR DESCRIPTION
**Pull request by jacek-izykowski**

### CODE REVIEW

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc [hunk 1]
```diff
@@ -11,9 +11,14 @@
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <sys/types.h>
-//WINDOWS #include <sys/ipc.h>
-//WINDOWS #include <sys/shm.h>
-//WINDOWS #include <asm/types.h>
+#if !defined(_WIN32)
```
Comment (katrybacka) [inline src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: I suggest you to use here ifndef like in the rest of the file

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc [hunk 2]
```diff
@@ -126,9 +141,8 @@ void KSyncFlowMemory::InitFlowMem() {
     }
     nl_free_client(cl);
 
+#if defined(__linux__)
```
Comment (katrybacka) [inline src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: I suggest you to use here ifdef like in the rest of the file

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc [hunk 3]
```diff
@@ -148,7 +162,7 @@ void KSyncFlowMemory::InitFlowMem() {
     }
 #endif
 
-#if 0 //WINDOWSFIX
+#if !defined(_WIN32)
```
Comment (katrybacka) [inline src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: I suggest you to use here ifndef like in the rest of the file

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc [hunk 4]
```diff
@@ -388,9 +405,9 @@ void KSyncFlowMemory::GetFlowTableSize() {
 }
 
 void KSyncFlowMemory::MapSharedMemory() {
-#if 0 //WINDOWSFIX
     GetFlowTableSize();
 
+#if !defined(_WIN32)
```
Comment (katrybacka) [inline src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: I suggest you to use here ifndef like in the rest of the file

**Review by katrybacka — CHANGES_REQUESTED** [review]

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc (hunk 1 above)
Comment (jacek-izykowski) [reply src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: Original file uses "#if defined" version (see lines: 6, 8).

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc (hunk 2 above)
Comment (jacek-izykowski) [reply src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: Original file uses "#if defined" version (see lines: 6, 8).

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc (hunk 3 above)
Comment (jacek-izykowski) [reply src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: Original file uses "#if defined" version (see lines: 6, 8).

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc (hunk 4 above)
Comment (jacek-izykowski) [reply src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: Original file uses "#if defined" version (see lines: 6, 8).

**Review by katrybacka — APPROVED** [review]

File: src/vnsw/agent/vrouter/ksync/ksync_init.cc [hunk 5]
```diff
@@ -92,31 +91,23 @@ void KSync::RegisterDBClients(DB *db) {
...
     VRouterInterfaceSnapshot();
-    // TODO: JW-882: uncomment the code
-    /* InitFlowMem(); */
+    InitFlowMem();
     ResetVRouter(true);
-
-    /* On Windows, vhost interface is created when Hyper-V switch is created */
-    #ifndef _WIN32
     if (create_vhost) {
+        // On Windows, vhost interface is created when Hyper-V switch is
```
Comment (sodar) [inline src/vnsw/agent/vrouter/ksync/ksync_init.cc]: I would revert this change, so that the whole condition is enclosed in compilation guard. This way we could explicitly show that Agent on Windows is not responsible for creating the vhost interface.

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc [hunk 6]
```diff
@@ -406,12 +423,38 @@ void KSyncFlowMemory::MapSharedMemory() {
             << "> : " << strerror(errno));
         assert(0);
     }
+#else
+    flow_table_ = GetFlowTableMemoryFromWindowsPipe();
```
Comment (sodar) [inline src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: This can be removed, because that code will be called only on FreeBSD.

**Review by sodar — DISMISSED** [review]

File: src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc (hunk 6 above)
Comment (jacek-izykowski) [reply src/vnsw/agent/vrouter/ksync/ksync_flow_memory.cc]: This is correct.
